### DISCUSSION TITLE
[v3][Forms] Call action function without rendering from custom field

### DISCUSSION THREAD

**Discussion by sprtk-ches** [filamentphp/filament#8865] [Help] [answered]
I'm trying to create a custom field and I want a way for the field to be able to change some internal state. 

For example let's say I have the following Component

```php
class OptionPicker extends Field
{
    protected ?Collection $options;

    protected function setUp(): void
    {
        $this->registerActions([
            Action::make('setOptions')
                ->action('changeOptions'),
        ]);
    }

    public function options(Collection $options): static
    {
        $this->options = $options;
        return $this;
    }

    public function getOptions(): Collection
    {
        return $this->evaluate($this->options);
    }

    protected function changeOptions()
    {
        $this->options = new Collection(['New Options']); // Just an example
    }
}

```

```html
<x-dynamic-component
    :component="$getFieldWrapperView()"
    :field="$field"
>
    <div x-data="{ state: $wire.$entangle('{{ $getStatePath() }}') }">
        @foreach($getOptions() as $option)
            <div @click="setOptions">Change options</div>  <-- For demo purposes only. This is what I want to achieve
        @endforeach
    </div>
</x-dynamic-component>
```

The above is a WIP how I tried to achieve this (by following the instructions [here](https://filamentphp.com/docs/3.x/forms/actions#adding-an-action-to-a-custom-form-component)).
However, I'm not sure **how to actually bind an action to a click event**. From what I've seen, the suggested way is to do something like
```
{{ $getAction('setOptions') }}
```

However this renders the button, which is something I don't want. I tried to pass a custom view to the action
```php
        $this->registerActions([
            Action::make('setOptions')
                ->view('my.custom.view')
                ->action('changeOptions'),
        ]);
```
however I'm not sure how to pass arguments to the this function/view (`viewData` doesn't work as I don't have the data when I define the view)

So, **is there a way to call/bind the action function in a click method without rendering it? And how would I pass arguments to it?** 

I'm a novice in both livewire and filament so I apologise if I'm missing something obvious. And sorry for the wall of text.

**Comment by leandrocfe** [comment]
maybe using `afterStateUpdated`?

```php
YourComponent::make('field')
    ->live()
    ->afterStateUpdated(function (?string $state) {
        //...
    })
```

```html
<x-dynamic-component :component="$getFieldWrapperView()" :field="$field">
    <div x-data="{ state: $wire.{{ $applyStateBindingModifiers("\$entangle('{$getStatePath()}')") }} }">
        <div>
            @foreach ($options as $key => $value)
                <div @click="state = {{ $key }}">
                    {{ $value }}</div>
            @endforeach
        </div>
    </div>
</x-dynamic-component>
```

**Reply by sprtk-ches** [reply]
🤔 This could work as a workaround but is tied to the state being changed. Ideally, I want to be able to control that, not rely on it (if that makes sense).

In any case, the best solution I found is to put the methods I want in the form component (eg `EditUser`). This way I have access to all Livewire tools. I would rather the form field was self-contained though. (TBF I have a pretty specific use-case so this is not a complaint)

Either way, thanks for your time and suggestion 🙇

**Reply by leandrocfe** [reply, marked as answer]
hum, maybe registering listeners?

```php
->registerListeners([
    'myComponent::updated' => [
        function (Component $component, string $statePath, string $customValue): void {
            //...
        }
    ]
])
```

```html
<button 
    type="button"
    wire:click="dispatchFormEvent('myComponent::updated', '{{ $getStatePath() }}', 'custom value')"
>update
</button>
```

**Reply by sprtk-ches** [reply]
I wasn't familiar with events, but I gave it a try and this might actually be what I'm looking for, TYVM! 🙇  I've already solved my issue with a workaround but hopefully it might help someone with the same problem.

PS: If anybody is new to Filament, you can register your event listeners in the `setUp()` function of the component. So something like this

```php
class MyCustomField extends Field
{
    protected function setUp(): void
    {
        $this->registerListeners([
            'testEvent::triggered' => [
                function ($component, $customValue) {
                    // Your code here
                },
            ],
        ]);
    }
}
```
